Add support for styling grammar and spelling errors

pxlcoder opened the pull request:
#### dea7937dec2bc7d25ea94641cb60bdf586f6526e
<pre>
Add support for styling grammar and spelling errors
<a href="https://bugs.webkit.org/show_bug.cgi?id=265879">https://bugs.webkit.org/show_bug.cgi?id=265879</a>
<a href="https://rdar.apple.com/119190841">rdar://119190841</a>

Reviewed by Antti Koivisto.

Support the use of `::grammar-error` and `::spelling-error` to change the color,
background color, fill color, stroke color, and text decoration of grammar and
spelling errors.

Similar to `::highlight()`, support for `text-shadow` and `stroke-width` is
currently missing. To adhere to the spec, support for these properties should
be added for all highlight pseudo-elements in subsequent changes.

Spec: <a href="https://www.w3.org/TR/css-pseudo-4/#highlight-styling">https://www.w3.org/TR/css-pseudo-4/#highlight-styling</a>

* LayoutTests/editing/spelling/grammar-and-spelling-error-styling-expected.html: Added.
* LayoutTests/editing/spelling/grammar-and-spelling-error-styling.html: Added.
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForDocumentMarkers):

Ensure grammar and spelling markers are collected for all paint phases, since
they are fully styleable.

* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::textSegmentPseudoStyle const):

Factor out pseudo-element style retrieval for text nodes into a common method.

Additionally, use `getCachedPseudoStyle` rather than `getUncachedPseudoStyle`,
since `::selection`, `::grammar-error`, and `::spelling-error` do not take
additional parameters.

(WebCore::RenderElement::selectionColor const):
(WebCore::RenderElement::selectionPseudoStyle const):
(WebCore::RenderElement::selectionBackgroundColor const):
(WebCore::RenderElement::spellingErrorPseudoStyle const):
(WebCore::RenderElement::grammarErrorPseudoStyle const):
* Source/WebCore/rendering/RenderElement.h:

Update comment describing when to use `getCachedPseudoStyle` and `getUncachedPseudoStyle`
for accuracy.

* Source/WebCore/rendering/RenderText.h:
(WebCore::RenderText::selectionPseudoStyle const):
(WebCore::RenderText::spellingErrorPseudoStyle const):
(WebCore::RenderText::grammarErrorPseudoStyle const):
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::computeStyleForPseudoElementStyle):

Factor out `StyledMarkedText` generation into a common method used by highlights
and the grammar and spelling pseudo elements.

(WebCore::resolveStyleForMarkedText):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintForegroundAndDecorations):

Detect text decorations when grammar and spelling pseudo styles are applied to
ensure the paint phase is not skipped.

(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintPlatformDocumentMarkers):

Do not paint platform error underlines if a custom text decoration is specified
using the new pseudo-elements.

* Source/WebCore/rendering/TextPaintStyle.cpp:
(WebCore::computeTextSelectionPaintStyle):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _setGrammarCheckingEnabledForTesting:]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setGrammarCheckingEnabled:]):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cocoaResetStateToConsistentValues):

Always enable grammar checking for testing. Grammar checking is already enabled
by default on most Cocoa platforms other than older macOS.

Canonical link: <a href="https://commits.webkit.org/271643@main">https://commits.webkit.org/271643@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04e1156bbb18dc7e7c4f1864822cd498204827b4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29094 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7763 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30433 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31695 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26483 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29690 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9888 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5069 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29366 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6413 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24943 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5566 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5689 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25973 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33033 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26569 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26409 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/31939 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5669 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3842 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7334 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/25769 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6945 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6173 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -1654,6 +1654,33 @@ std::unique_ptr<RenderStyle> RenderElement::getUncachedPseudoStyle(const Style::
+    if (isAnonymous())
+        return nullptr;
+
+    if (auto pseudoStyle = getUncachedPseudoStyle({ pseudo })) {
```
Why not use getCachedPseudoStyle? The reason we can't for highlight is because of the parameter name ::highlight(...)

pxlcoder replied:
I was following what's already done for `::selection`.

We have the following comment above the methods:
```
// The pseudo element style can be cached or uncached.  Use the cached method if the pseudo element doesn't respect
// any pseudo classes (and therefore has no concept of changing state).
```

Though from testing, I don't think that `::selection` works with other pseudo classes either? Maybe @anttijk can weigh in on what the right thing to do here is.

anttijk replied:
I don't know what that comment means. getCachedPseudoStyle is the right choice if it works. getUncachedPseudoStyle should only be needed if you have to pass additional parameters (like `::highlight()` name).

pxlcoder replied:
Updated to use `getCachedPseudoStyle` after testing it works (tested with `::selection`, in addition to the new pseudo-elements).

pxlcoder committed:
Add support for styling grammar and spelling errors
https://bugs.webkit.org/show_bug.cgi?id=265879
rdar://119190841

Reviewed by Antti Koivisto.

Support the use of `::grammar-error` and `::spelling-error` to change the color,
background color, fill color, stroke color, and text decoration of grammar and
spelling errors.

Similar to `::highlight()`, support for `text-shadow` and `stroke-width` is
currently missing. To adhere to the spec, support for these properties should
be added for all highlight pseudo-elements in subsequent changes.

Spec: https://www.w3.org/TR/css-pseudo-4/#highlight-styling

* LayoutTests/editing/spelling/grammar-and-spelling-error-styling-expected.html: Added.
* LayoutTests/editing/spelling/grammar-and-spelling-error-styling.html: Added.
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForDocumentMarkers):

Ensure grammar and spelling markers are collected for all paint phases, since
they are fully styleable.

* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::textSegmentPseudoStyle const):

Factor out pseudo-element style retrieval for text nodes into a common method.

Additionally, use `getCachedPseudoStyle` rather than `getUncachedPseudoStyle`,
since `::selection`, `::grammar-error`, and `::spelling-error` do not take
additional parameters.

(WebCore::RenderElement::selectionColor const):
(WebCore::RenderElement::selectionPseudoStyle const):
(WebCore::RenderElement::selectionBackgroundColor const):
(WebCore::RenderElement::spellingErrorPseudoStyle const):
(WebCore::RenderElement::grammarErrorPseudoStyle const):
* Source/WebCore/rendering/RenderElement.h:

Update comment describing when to use `getCachedPseudoStyle` and `getUncachedPseudoStyle`
for accuracy.

* Source/WebCore/rendering/RenderText.h:
(WebCore::RenderText::selectionPseudoStyle const):
(WebCore::RenderText::spellingErrorPseudoStyle const):
(WebCore::RenderText::grammarErrorPseudoStyle const):
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::computeStyleForPseudoElementStyle):

Factor out `StyledMarkedText` generation into a common method used by highlights
and the grammar and spelling pseudo elements.

(WebCore::resolveStyleForMarkedText):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter<TextBoxPath>::paintForegroundAndDecorations):

Detect text decorations when grammar and spelling pseudo styles are applied to
ensure the paint phase is not skipped.

(WebCore::TextBoxPainter<TextBoxPath>::paintPlatformDocumentMarkers):

Do not paint platform error underlines if a custom text decoration is specified
using the new pseudo-elements.

* Source/WebCore/rendering/TextPaintStyle.cpp:
(WebCore::computeTextSelectionPaintStyle):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _setGrammarCheckingEnabledForTesting:]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setGrammarCheckingEnabled:]):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cocoaResetStateToConsistentValues):

Always enable grammar checking for testing. Grammar checking is already enabled
by default on most Cocoa platforms other than older macOS.

Canonical link: https://commits.webkit.org/271643@main
